manager/src/main/java/org/openremote/manager/rules/AssetQueryPredicate.java

```diff
@@ -175,8 +178,28 @@ public static Predicate<NameValueHolder<?>> asPredicate(Supplier<Long> currentMi
             });
         }
 
-        return nameValueHolder -> namePredicate.test(nameValueHolder.getName())
-            && valuePredicate.test(valueExtractor.get().apply(nameValueHolder));
+        AtomicReference<Map<String, Long>> matchTimestamps = new AtomicReference<>(new HashMap<>());
```

Comment:
Added a custom asAttributeMatcher  as discussed, which is added within the JsonRulesBuilder. Also I reset the AssetQueryPredicate to its upstream state since no changes were needed there anymore. 

The duration handling is then done after the custom assetPredicate is applied and the matches are returned.